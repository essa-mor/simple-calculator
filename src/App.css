.App {
  display: inline-grid;
  grid-template-columns: auto auto auto auto;
  padding: 1px;
  grid-column-gap: 1px;
  grid-row-gap: 1px;
  background-color: rgb(146, 16, 41);
  width: 164px;
}

.result{
  grid-column-start: 1;
  grid-column-end: 5;
}

.digit, .clear_result{
  color: rgb(42, 42, 42);
  background-color:white;
  border: none;
  padding: 0;
  margin: 0;
  height: 36px;
}

.clear_result{
  grid-column-start: 1;
  grid-column-end: 3;
}

.operator, .equal{
  color: white;
  background-color: rgb(242, 39, 79);
  border: none;
  padding: 0;
  margin: 0;
  height: 36px;
}

.equal{
  background-color: rgb(202, 30, 39);
  grid-column-start: 3;
  grid-column-end: 5;
}