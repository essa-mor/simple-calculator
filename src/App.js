import React, { Component } from 'react';

import ResultPanel from './components/ResultPanel';
import './App.css';

const initialState = { prevResult: 0, result: 0, operator: '', summary: '' };

class App extends Component {
	constructor(props) {
		super(props);
		this.state = initialState;
		this.onDigitClick = this.onDigitClick.bind(this);
		this.onClearClick = this.onClearClick.bind(this);
		this.onKeyPress = this.onKeyPress.bind(this);
		this.onOperatorClick = this.onOperatorClick.bind(this);
		this.onEqualClick = this.onEqualClick.bind(this);
	}

	onClearClick() {
		this.setState(initialState);
	}

	onDigitClick(e) {
		const { value } = e.target;
		this.addDigit(value);
	}

	addDigit(value) {
		this.setState(prevState => {
			const result = prevState.operator !== '' ? 0 : prevState.result;
			return {
				result: result * 10 + parseInt(value)
			};
		});
	}

	onKeyPress(e) {
		if (e.key >= 0 && e.key <= 9) {
			this.addDigit(e.key);
		} else if (['-', '+', '/', '*'].includes(e.key)) {
			this.setOperator(e.key);
		} else if (e.key === '=' /*|| e.key === 'Enter'*/) {
			this.setResult();
		}
	}

	onOperatorClick(e) {
		const { value } = e.target;
		this.setOperator(value);
	}

	setOperator(operator) {
		this.setState(prevState => ({
			operator,
			summary: `${prevState.summary} ${prevState.result} ${operator}`,
			prevResult: prevState.result
		}));
	}

	onEqualClick() {
		this.setResult();
	}

	setResult() {
		this.setState(prevState => ({
			...initialState, result: eval(`${prevState.prevResult}  ${prevState.operator}  ${prevState.result}`)
		}));
	}

	render() {
		const { result, summary } = this.state;
		return (
			<div className="App" onKeyPress={this.onKeyPress}>
				<ResultPanel className="result" summary={summary} result={result}/>
				<button type="button" onClick={this.onClearClick} className="clear_result">Clear</button>
				<button type="button" value="-" onClick={this.onOperatorClick} className="operator">-</button>
				<button type="button" value="7" onClick={this.onDigitClick} className="digit">7</button>
				<button type="button" value="8" onClick={this.onDigitClick} className="digit">8</button>
				<button type="button" value="9" onClick={this.onDigitClick} className="digit">9</button>
				<button type="button" value="/" onClick={this.onOperatorClick} className="operator">/</button>
				<button type="button" value="4" onClick={this.onDigitClick} className="digit">4</button>
				<button type="button" value="5" onClick={this.onDigitClick} className="digit">5</button>
				<button type="button" value="6" onClick={this.onDigitClick} className="digit">6</button>
				<button type="button" value="*" onClick={this.onOperatorClick} className="operator">*</button>
				<button type="button" value="1" onClick={this.onDigitClick} className="digit">1</button>
				<button type="button" value="2" onClick={this.onDigitClick} className="digit">2</button>
				<button type="button" value="3" onClick={this.onDigitClick} className="digit">3</button>
				<button type="button" value="+" onClick={this.onOperatorClick} className="operator">+</button>
				<button type="button" value="0" onClick={this.onDigitClick} className="digit">0</button>
				<button type="button" className="digit">.</button>
				<button type="button" onClick={this.onEqualClick} className="equal">=</button>
			</div>
		);
	}
}

export default App;
