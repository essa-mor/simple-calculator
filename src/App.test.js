import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App from './App';

describe('App test', () => {
	it('renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(<App />, div);
		ReactDOM.unmountComponentAtNode(div);
	});	

	it('renders properly', () => {
		const tree = renderer.create(<App />).toJSON();
		expect(tree).toMatchSnapshot();
	});	
});
