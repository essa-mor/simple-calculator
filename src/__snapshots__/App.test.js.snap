// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App test renders properly 1`] = `
<div
  className="App"
>
  <div
    className="result"
  >
    <span
      className="result_label"
    >
      
    </span>
    <input
      className="result_input"
      readOnly={true}
      type="text"
      value="0"
    />
  </div>
  <button
    className="clear_result"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="Clear"
  >
    Clear
  </button>
  <button
    className="digit"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="←"
  >
    ←
  </button>
  <button
    className="operator"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="-"
  >
    -
  </button>
  <button
    className="digit"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="7"
  >
    7
  </button>
  <button
    className="digit"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="8"
  >
    8
  </button>
  <button
    className="digit"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="9"
  >
    9
  </button>
  <button
    className="operator"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="/"
  >
    /
  </button>
  <button
    className="digit"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="4"
  >
    4
  </button>
  <button
    className="digit"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="5"
  >
    5
  </button>
  <button
    className="digit"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="6"
  >
    6
  </button>
  <button
    className="operator"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="*"
  >
    *
  </button>
  <button
    className="digit"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="1"
  >
    1
  </button>
  <button
    className="digit"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="2"
  >
    2
  </button>
  <button
    className="digit"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="3"
  >
    3
  </button>
  <button
    className="operator"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="+"
  >
    +
  </button>
  <button
    className="digit"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="0"
  >
    0
  </button>
  <button
    className="digit"
    onClick={[Function]}
    onFocus={[Function]}
    type="button"
    value="."
  >
    .
  </button>
  <button
    className="equal"
    onClick={[Function]}
    type="button"
  >
    =
  </button>
</div>
`;
